guacamole/src/main/java/org/apache/guacamole/rest/connection/ConnectionResource.java

```diff
@@ -150,18 +149,16 @@ public ConnectionResource(@Assisted UserContext userContext,
      * @throws GuacamoleException
      *     If an error occurs while retrieving the connection history.
      */
-    @GET
     @Path("history")
-    public List<APIConnectionRecord> getConnectionHistory()
+    public ConnectionHistoryResource getConnectionHistory()
             throws GuacamoleException {
 
-        // Retrieve the requested connection's history
-        List<APIConnectionRecord> apiRecords = new ArrayList<APIConnectionRecord>();
-        for (ConnectionRecord record : connection.getHistory())
-            apiRecords.add(new APIConnectionRecord(record));
-
-        // Return the converted history
-        return apiRecords;
+        try {
+            return new ConnectionHistoryResource(connection.getConnectionHistory());
+        }
+        catch (GuacamoleUnsupportedException e) {
+            return new ConnectionHistoryResource(new SimpleActivityRecordSet<>());
+        }
```

Comment:
Yeah, I think I'm finally starting to get my brain around handling all of the compatibility across implementations. I see what you're saying, and I'll take a shot at tweaking it to your suggestion, here, and punching up the documentation to cover it.

I _think_ that covers all the scenarios, but I'll also try to go over possibilities in my mind and make sure there aren't any corner cases that haven't been considered that would fail in such a scenario.

Reply:
Okay, I've worked through this - not entirely sure how well I've done. I hit some issues with the `List<? extends ConnectionRecord>` and converting that to `List<ConnectionRecord>`. I ended up with the loops you see, here, now - not sure if there's a more elegant way to do that translation from `? extends RecordType` to just `RecordType`? I tried various adjustments to the types of methods and such in the classes, and I couldn't really do much without breaking compatibility.

Also, the `HistoryResource` class in `rest/history` pulls history from `UserContext`, whereas the `ConnectionResource` and `UserResource` classes in `rest/connection` and `rest/user` pull history from `Connection` and `User`, and I'm not sure if there are any modifications needed to the `HistoryResource` class or anything in the `UserContext` interfaces or classes that need to be tweaked to handle these changes?

Reply:
> Okay, I've worked through this - not entirely sure how well I've done. I hit some issues with the `List<? extends ConnectionRecord>` and converting that to `List<ConnectionRecord>`. I ended up with the loops you see, here, now - not sure if there's a more elegant way to do that translation from `? extends RecordType` to just `RecordType`? ...

Can you point to the specific part of the code where you're encountering trouble? I don't think there should be any need to convert.

> Also, the `HistoryResource` class in `rest/history` pulls history from `UserContext`, whereas the `ConnectionResource` and `UserResource` classes in `rest/connection` and `rest/user` pull history from `Connection` and `User`, and I'm not sure if there are any modifications needed to the `HistoryResource` class or anything in the `UserContext` interfaces or classes that need to be tweaked to handle these changes?

Are you referring to whether the code that pulls directly from `UserContext` needs to add special handling for `GuacamoleUnsupportedException`?

Reply:
> Can you point to the specific part of the code where you're encountering trouble? I don't think there should be any need to convert.

See comment below: https://github.com/apache/guacamole-client/pull/546#discussion_r513709095

> Are you referring to whether the code that pulls directly from UserContext needs to add special handling for GuacamoleUnsupportedException?

Correct, or whether `UserContext` itself needs to take that into account at all.

Reply:
I agree that it would make sense to be consistent in the handling of `GuacamoleUnsupportedException` and provide the same functionality there, as well. It could make things easier for developers that may well use the same code to drive both.